torchbenchmark/models/dcgan/__init__.py

```diff
@@ -0,0 +1,334 @@
+        if (dcgan.device == 'cuda') and (ngpu > 1):
+            self.netD = nn.DataParallel(netD, list(range(ngpu)))
+
+        # Apply the weights_init function to randomly initialize all weights
+        #  to mean=0, stdev=0.2.
+        self.netD.apply(weights_init)
+        
+        if self.debug_print:
+            # Print the model
+            print(self.netD)
+
+        # Initialize BCELoss function
+        self.criterion = nn.BCELoss()
+
+        # Create batch of latent vectors that we will use to visualize
+        #  the progression of the generator
+        self.fixed_noise = torch.randn(64, nz, 1, 1, device=device)
+
+        # Establish convention for real and fake labels during training
+        self.real_label = 1.
+        self.fake_label = 0.
+
+        # Setup Adam optimizers for both G and D
+        self.optimizerD = optim.Adam(self.netD.parameters(), lr=lr, betas=(beta1, 0.999))
+        self.optimizerG = optim.Adam(self.netG.parameters(), lr=lr, betas=(beta1, 0.999))
+
+        # Train batch size
+        self.train_b_size = 32
+
+        # inference would just run descriminator so thats what we'll do too.
+        self.inference_just_descriminator = True
+
+        # eval batch size
+        self.eval_b_size = 256
+
+        self.eval_fake = torch.randn(self.eval_b_size, 3, 64, 64, device=self.device)
+
+        if self.jit:
+            self.netD = torch.jit.trace(self.netD,(self.eval_fake,))
+
+        if False == self.inference_just_descriminator:
+            self.eval_noise = torch.randn(self.eval_b_size, nz, 1, 1, device=self.device)
+            if self.jit:
+              self.netG = torch.jit.trace(self.netG,(self.eval_noise,))
+
+    def get_module(self):
+        return self.netD, (self.eval_fake,)
+
+    def eval(self, niter=1):
+
+        for _ in range(niter):
+
+            if False == self.inference_just_descriminator:
+              # Generate fake image batch with G
+              self.eval_fake = self.netG(self.eval_noise)
+
+            # Since we just updated D, perform another forward pass of all-fake batch through D
+            output = self.netD(self.eval_fake).view(-1)
+
+        return
+
+    def train(self, niter=1):
+
+        # Training Loop
+
+        # Lists to keep track of progress
+        img_list = []
+        iters = 0
+
+        dcgan = self.dcgan
+        device = dcgan.device
+
+        num_epochs = dcgan.num_epochs
+        num_train_batch = niter
+
+        lr = dcgan.lr
+        nz = dcgan.nz
+        beta1 = dcgan.beta1
+
+        netD = self.netD
+        netG = self.netG
+
+        criterion = self.criterion
+        optimizerD = self.optimizerD
+        optimizerG = self.optimizerG
+
+        real_label = self.real_label
+        fake_label = self.fake_label
+
+        # Random values as surrogate for batch of photos
+        benchmark_pic = torch.randn(self.train_b_size, 3, 64, 64, device=self.device)
+
+        # For each epoch
+        for epoch in range(num_epochs):
```

Comment:
Here we are fix `num_epochs` to 5 and setting `num_train_batch` to `niter`. I am thinking whether it is needed to run 5 epochs. Maybe set num_epochs to `niter` and `num_train_batch` to 1, since we probably want to run only 1 batch?

Reply:
It could be. I only uploaded this so I could profile on my other machine. The dcgan is crossing two models during training and there's significant cpu time between each epoch. I want to be sure to capture the damage between teh epochs but count need to be balanced against runtime.